cmd/config/docs/api-conventions/functions-spec.md

```diff
@@ -18,168 +23,372 @@ interpreted as described in [RFC 2119][2].
-to the function.
+The following is an example output containing one result representing a
+validation error:
 
-### Output Type
-
-A function’s output MUST be the same as the input specification above
--- i.e. `ResourceList` or `List`.
-This is necessary to enable chaining two or more functions together in a pipeline.
-The serialization format of the output SHOULD match that of its input on each invocation
--- e.g. if the input was a `ResourceList`, the output should also be a `ResourceList`.
+```yaml
+apiVersion: config.kubernetes.io/v1
+kind: ResourceList
+items:
+  - apiVersion: v1
+    kind: Service
+    metadata:
+      name: wordpress
+      labels:
+        app: wordpress
+      annotations:
+        config.kubernetes.io/index: "0"
+        config.kubernetes.io/path: "service.yaml"
+    spec: # Example comment
+      type: LoadBalancer
+      selector:
+        app: wordpress
+        tier: frontend
+      ports:
+        - protocol: TCP
+          port: 80
+results:
+  - message: "Invalid type. Expected: integer, given: string"
+    severity: error
+    resourceRef:
+      apiVersion: v1
+      kind: Service
+      name: wordpress
+    field:
+      path: spec.ports.0.port
+    file:
+      path: service.yaml
+```
 
 ### Serialization Format
 
 A function MUST support YAML as a serialization format for the input and output.
-A function MUST use utf8 encoding (as YAML is a superset of JSON, JSON will also be supported
-by any conforming function).
+A function MUST use utf8 encoding (as YAML is a superset of JSON, JSON will also
+be supported by any conforming function).
+
+### Containerization
+
+A function MUST be implemented as a container.
+
+A function container MUST be capable of running as a non-root user `nobody` if
+it does not require access to host filesystem or makes network calls.
+
+### stderr
+
+Any non-structured error messages MUST be emitted to `stderr`. `stdout` is
+reserved for `ResourceList` as described above.
+
+### Exit Code
+
+An exit code of zero indicates function execution was successful. A non-zero
+exit code indicates a failure.
 
 ### Operations
 
-A function MAY Create, Update, or Delete any number of items in the `items` field and output the
-resultant list.
+A function MAY Create, Update, or Delete any number of items in the `items`
+field and output the resultant list in the corresponding `items` field of the
+output.
 
-A function MAY modify annotations with prefix `config.kubernetes.io`, but must be careful about
-doing so since they’re used for orchestration purposes and will likely impact subsequent functions
-in the pipeline.
+A function MAY modify annotations with prefix `config.kubernetes.io`, but must
+be careful about doing so since they’re used for orchestration purposes and will
+likely impact subsequent functions in the pipeline. Several annotations and
+their semantics are listed below.
 
 A function SHOULD preserve comments when input serialization format is YAML.
```

Comment:
SHOULD or MUST?

Reply:
SHOULD. In general, functions may be implemented in toolchains that wont' preserve comments. The orchestrator is/should be smart enough to reconcile the comments.